torch/csrc/jit/codegen/cuda/runtime/warp.cu

```diff
@@ -0,0 +1,55 @@
+namespace warp {
+
+const int WARP_SIZE = 32;
+
+template <
+    bool SINGLE_WARP,
+    typename T,
+    typename Func,
+    typename _dim3ti,
+    typename _dim3bd>
+__device__ void fusedBlockReduceBroadcastX(
+    T& out,
+    const T& inp_val,
+    Func reduction_op,
+    const _dim3ti& thread_idx,
+    const _dim3bd& block_dim,
+    T* shared_mem,
+    bool read_write_pred,
+    T init_val) {
+  // Assume input padded to multiples of a warp
+  unsigned int linear_tid = thread_idx.x;
+  T reduce_val = init_val;
+
+  // Load value to shared mem
+  if (!SINGLE_WARP) {
```

Comment:
I would like to see the structure here move from shared memory -> warp shuffle to something like:
Reduce all warps -> write to SMEM
First warp reads from SMEM and finishes reduction (can be in smem or shuffle reduction)

Reply:
I was trying to make the fused reduction and broadcast work for the cases with more than one warp. I'm not yet sure if there's a better way to support this if we warp reduce first.